x-pack/plugins/security_solution/public/common/containers/source/index.tsx

```diff
@@ -275,20 +310,43 @@ export const useIndexFields = (sourcererScopeName: SourcererScopeName) => {
             },
           });
       };
+
       searchSubscription$.current.unsubscribe();
       abortCtrl.current.abort();
       asyncSearch();
     },
-    [data.search, dispatch, addError, addWarning, setLoading, sourcererScopeName]
+    [
+      addError,
+      addWarning,
+      data.search,
+      dispatch,
+      patternList,
+      selectedPatterns,
+      setLoading,
+      signalIndexNameSelector,
+      sourcererScopeName,
+    ]
   );
+  const refDataViewId = useRef('');
+  const refSelectedPatterns = useRef([] as string[]);
 
   useEffect(() => {
-    if (!isEmpty(indexNames) && previousIndexNames !== indexNames.sort().join()) {
-      indexFieldsSearch(indexNames);
+    if (
+      (dataViewId != null && dataViewId !== refDataViewId.current && selectedPatterns.length > 0) ||
+      (selectedPatterns.length > 0 && refSelectedPatterns.current.length === 0)
+    ) {
+      indexFieldsSearch(dataViewId);
     }
+    refSelectedPatterns.current = selectedPatterns;
+    refDataViewId.current = dataViewId;
+  }, [dataViewId, indexFieldsSearch, selectedPatterns]);
+
+  useEffect(() => {
     return () => {
       searchSubscription$.current.unsubscribe();
       abortCtrl.current.abort();
     };
-  }, [indexNames, indexFieldsSearch, previousIndexNames]);
+  }, []);
+
+  return { indexFieldsSearch };
```

Comment:
need to refetch fields the first time signals index is created

Reply:
I think here we can do an Object.freeze, but I still need to understand why we need to do it like that and not just passing the signal index name as a props of the attributes. I think we already have a hook [`useFetchIndex`](https://github.com/elastic/kibana/blob/master/x-pack/plugins/security_solution/public/common/containers/source/index.tsx#L125) which just take an index name and get what it is appropriate